File: ios/Survolocking/Engine/PhoneUtils.swift
import Foundation
import CryptoKit

/**
 * 号码处理。
 *
 * 端云哈希契约：必须与 Android 端 PhoneUtils、服务端 security.hash_phone 完全一致，
 * 三端统一为 HMAC-SHA256(salt, normalize(phone))。
 *
 * 盐的来源：服务端在登录 / me 时下发 PHONE_HASH_SALT，端侧缓存到 Keychain/UserDefaults
 * （salt 变量），作为哈希计算的唯一真相源；内置 "survolocking_phone_salt_v1" 仅为
 * 登录前尚未拿到服务端盐时的兜底默认值，真实部署会被覆盖。
 * 否则同一号码在 Android 上传的日志，iOS 无法匹配服务端下发的规则。
 */
enum PhoneUtils {

    /// 当前生效的盐，默认取兜底常量，登录后被服务端下发值覆盖
    private static var _salt = "survolocking_phone_salt_v1"
    static var salt: String { _salt }

    /// 用服务端下发的盐更新，空值忽略
    static func setSalt(_ salt: String) {
        if !salt.isEmpty { _salt = salt }
    }

    /// 归一化：去除空格、横线、括号，保留前导 +
    static func normalize(_ phone: String) -> String {
        let allowed = Set("0123456789+")
        return String(phone.filter { allowed.contains($0) })
    }

    static func hash(_ normalizedPhone: String) -> String {
        let key = SymmetricKey(data: Data(salt.utf8))
        let mac = HMAC<SHA256>.authenticationCode(
            for: Data(normalizedPhone.utf8), using: key
        )
        return mac.map { String(format: "%02x", $0) }.joined()
    }

    static func hash(of rawPhone: String) -> String {
        hash(normalize(rawPhone))
    }

    /// 异常号段：境外来电、虚拟运营商、高仿号
    static func isSuspiciousPrefix(_ normalized: String) -> Bool {
        var n = normalized
        if n.hasPrefix("+") { n.removeFirst() }
        let prefixes = [
            "00852", "00853", "00886",           // 中国香港 / 中国澳门 / 中国台湾
            "001", "0060", "0065", "0081", "0082", // 常见境外诈骗来源
            "170", "171", "162", "165", "167", "174", "149" // 虚拟运营商
        ]
        return prefixes.contains { n.hasPrefix($0) }
    }

    /// 展示用脱敏号码
    static func masked(_ raw: String) -> String {
        let n = normalize(raw)
        guard n.count >= 7 else { return "未知号码" }
        let head = n.prefix(3)
        let tail = n.suffix(4)
        return "\(head)****\(tail)"
    }
}

File: android/app/src/main/kotlin/com/survolocking/engine/PhoneUtils.kt
package com.survolocking.engine

import com.survolocking.data.Constants
import java.security.MessageDigest
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * 号码处理工具。
 *
 * 端云哈希契约：此处算法必须与服务端 security.hash_phone 完全一致，
 * 即 HMAC-SHA256(activeSalt, normalized)。
 *
 * 盐的来源：服务端在登录 / /me 时下发 PHONE_HASH_SALT，端侧缓存到本地
 * （activeSalt），作为哈希计算的唯一真相源；Constants.PHONE_SALT 仅作为
 * 登录前尚未拿到服务端盐时的兜底默认值。这样无需手动同步三端源码常量，
 * 避免盐不一致导致日志与规则静默失配。
 */
object PhoneUtils {

    private val NON_DIGIT_KEEP_PLUS = Regex("[^0-9+]")

    /** 当前生效的盐，默认取兜底常量，登录后被服务端下发值覆盖 */
    var activeSalt: String = Constants.PHONE_SALT
        private set

    /** 用服务端下发的盐更新，空值忽略 */
    fun setSalt(salt: String) {
        if (salt.isNotBlank()) activeSalt = salt
    }

    /** 归一化：去空格、横线、括号，保留前导 + */
    fun normalize(phone: String): String =
        NON_DIGIT_KEEP_PLUS.replace(phone, "")

    fun hash(normalizedPhone: String): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(activeSalt.toByteArray(), "HmacSHA256"))
        return mac.doFinal(normalizedPhone.toByteArray()).toHex()
    }

    fun hashOf(rawPhone: String): String = hash(normalize(rawPhone))

    /** 不依赖盐的裸哈希，仅用于本地临时索引，禁止上传 */
    fun sha256(input: String): String =
        MessageDigest.getInstance("SHA-256").digest(input.toByteArray()).toHex()

    private fun ByteArray.toHex(): String =
        joinToString("") { "%02x".format(it) }

    /**
     * 异常号段判定。
     * 境外来电、虚拟运营商号段、高仿号为骚扰电话高发区。
     */
    fun isSuspiciousPrefix(normalized: String): Boolean {
        val n = normalized.removePrefix("+")
        val prefixes = arrayOf(
            "00852", "00853", "00886", // 中国香港/澳门/台湾
            "001", "0060", "0065", "0081", "0082", // 常见境外诈骗来源
            "170", "171", "162", "165", "167", "174", "149" // 虚拟运营商
        )
        return prefixes.any { n.startsWith(it) }
    }
}
